src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -61,20 +64,35 @@ public void run() {
    */
   private void runNodes(List<T> freeNodes) {
     List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
+    CountDownLatch latch = null;
+    if (useLatches && runnables.size() > 1) {
+      latch = new CountDownLatch(runnables.size());
+    }
+    boolean latchUsed = false;
     for (IWorker<T> r : runnables) {
       m_activeRunnables.add(r);
+      if (latch != null && (r instanceof TestMethodWorker)) {
+        //Being conservative here and setting this only for test method workers
+        r.setLatch(latch);
+        latchUsed = true;
+      }
       ppp("Added to active runnable");
       setStatus(r, Status.RUNNING);
       ppp("Executing: " + r);
       try {
         execute(r);
-//        if (m_threadCount > 1) execute(r);
-//        else r.run();
       }
       catch(Exception ex) {
         ex.printStackTrace();
       }
     }
+    if (latch != null && latchUsed) {
+      try {
+        latch.await(10, TimeUnit.SECONDS);
```

Comment:
Ok, and hat about a less intrusive change like:

```java
  private void runNodes(List<T> freeNodes) {
    List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
    m_activeRunnables.addAll(runnables);
    List<Callable<Void>> callables = Lists.newArrayList(runnables.size());
    for (final IWorker<T> r : runnables) {
      setStatus(r, Status.RUNNING);
      callables.add(new Callable<Void>() {
        @Override
        public Void call() throws Exception {
          r.run();
          return null;
        }
      });
    }
    try {
      List<Future<Void>> futures = invokeAll(callables);
      for (Future<Void> future : futures) {
        try {
          future.get(10, TimeUnit.SECONDS);
        } catch (ExecutionException | TimeoutException e) {
          e.printStackTrace();
        }
      }
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }
```
?

BTW, I hope the second iteration is:

```
TestNG1.test2TestNG1()[pri:2, instance:test.github1336.TestNG1@23ab930d]
TestNG2.test2TestNG2()[pri:2, instance:test.github1336.TestNG2@4534b60d]
TestNG3.test2TestNG3()[pri:2, instance:test.github1336.TestNG3@3fa77460]
```

And the third one is:

```
TestNG1.test3TestNG1()[pri:3, instance:test.github1336.TestNG1@23ab930d]
TestNG2.test3TestNG2()[pri:3, instance:test.github1336.TestNG2@4534b60d]
TestNG3.test3TestNG3()[pri:3, instance:test.github1336.TestNG3@3fa77460]
```

(or we have an issue somewhere)